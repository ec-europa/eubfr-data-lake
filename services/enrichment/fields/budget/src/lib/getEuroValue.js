// @flow
import request from 'request-promise-native';
import type { BudgetItem, TimePrecision } from '@eubfr/types';

const padDate = (date: number): number | string =>
  date < 10 ? `0${date}` : date;

export const getEuroValue = async (
  inputBudgetItem: BudgetItem,
  projectEndDate: Date,
  projectEndPrecision: TimePrecision
) => {
  let euroValue = null;
  const precision = 100; // round converted value to the nearest 100

  // Least precision: year
  let period = projectEndDate.getFullYear();
  let apiPeriod = 'A';

  // Best precision: day
  if (projectEndPrecision === 'day') {
    const month = projectEndDate.getMonth() + 1;
    const day = projectEndDate.getDate();
    period = `${period}-${padDate(month)}-${padDate(day)}`;
    apiPeriod = 'D';
  }

  // If not a day, try at least: month
  else if (projectEndPrecision === 'month') {
    const month = projectEndDate.getMonth() + 1;
    period = `${period}-${padDate(month)}`;
    apiPeriod = 'M';
  }

  // "EXR/A.[currency].EUR.SP00.A" corresponds to the key of a dataset
  // The keys can be found here: http://sdw.ecb.europa.eu/browseTable.do?node=1495
  // API info: https://sdw-wsrest.ecb.europa.eu/web/generator/index.html
  const url = `https://sdw-wsrest.ecb.europa.eu/service/data/EXR/${apiPeriod}.${
    inputBudgetItem.currency
  }.EUR.SP00.A`;

  let results;
  const qs = {
    startPeriod: period,
    endPeriod: period,
    detail: 'dataonly',
  };

  try {
    results = await request.get({
      url,
      qs,
      json: true,
      headers: {
        Accept: 'application/vnd.sdmx.data+json;version=1.0.0-wd',
      },
    });
  } catch (e) {
    console.error('error', url, qs, e);
    return euroValue;
  }

  if (
    results &&
    results.dataSets &&
    results.dataSets[0] &&
    results.dataSets[0].series &&
    results.dataSets[0].series['0:0:0:0:0'] &&
    results.dataSets[0].series['0:0:0:0:0'].observations &&
    results.dataSets[0].series['0:0:0:0:0'].observations[0]
  ) {
    const exr = results.dataSets[0].series['0:0:0:0:0'].observations['0'][0];
    euroValue = Math.ceil(inputBudgetItem.value / exr / precision) * precision;
  }

  // Fallback request decreasing precision from daily to monthly
  if (results === undefined && projectEndPrecision === 'day') {
    euroValue = await getEuroValue(inputBudgetItem, projectEndDate, 'month');
  }

  return euroValue;
};

export default getEuroValue;
