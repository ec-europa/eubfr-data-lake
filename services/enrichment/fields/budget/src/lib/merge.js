/* eslint-disable no-underscore-dangle */
export const mergeRecords = (newRecord, existingRecord) => {
  if (!newRecord) return null;

  const mergedRecord = JSON.parse(JSON.stringify(newRecord));

  // What we need to retrieve from the existing record
  mergedRecord.computed_key = existingRecord.computed_key;

  // Budget: copy existing record if already enriched
  const budgetFields = [
    'eu_contrib',
    'other_contrib',
    'private_fund',
    'public_fund',
    'total_cost',
  ];

  if (existingRecord.budget && newRecord.budget) {
    budgetFields.forEach(field => {
      if (
        existingRecord.budget[field] &&
        mergedRecord.budget[field] &&
        existingRecord.budget[field]._original &&
        existingRecord.budget[field]._original.currency ===
          mergedRecord.budget[field].currency &&
        existingRecord.budget[field]._original.value ===
          mergedRecord.budget[field].value
      ) {
        mergedRecord.budget[field] = JSON.parse(
          JSON.stringify(existingRecord.budget[field])
        );
      }
    });
  }

  return mergedRecord;
};

export default mergeRecords;
