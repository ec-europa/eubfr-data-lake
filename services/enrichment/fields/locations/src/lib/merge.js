export const mergeRecords = (newRecord, existingRecord) => {
  if (!newRecord) return null;

  const mergedRecord = JSON.parse(JSON.stringify(newRecord));

  if (existingRecord) {
    // extract enriched data if available
  }

  return mergedRecord;
};

export default mergeRecords;
