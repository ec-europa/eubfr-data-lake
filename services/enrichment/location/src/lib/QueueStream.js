import stream from 'stream';

export default class QueueStream extends stream.Writable {
  constructor(options) {
    super(options);
    this.sqs = options.sqs;
    this.queueUrl = options.queueUrl;
  }

  _write(chunk, enc, next) {
    return this.sqs.sendMessage(
      {
        MessageAttributes: {},
        MessageBody: JSON.stringify(chunk),
        QueueUrl: this.queueUrl,
      },
      err => {
        if (err) {
          console.log('Error', err);
        }

        next();
      }
    );
  }
}
