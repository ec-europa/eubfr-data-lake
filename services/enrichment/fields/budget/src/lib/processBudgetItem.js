// @flow
import sanitizeBudgetItem from '@eubfr/lib/budgetFormatter';
import type { BudgetItem } from '@eubfr/types';

import { getEuroValue } from './getEuroValue';

// List of currencies that can be translated into EUR
// Manually extracted form ECB's API
// http://sdw.ecb.europa.eu/browseTable.do?node=1495 ("Currency" field)
const availableCurrencies = [
  'AUD',
  'BGN',
  'BRL',
  'CAD',
  'CHF',
  'CNY',
  'CYP',
  'CZK',
  'DKK',
  'DZD',
  'EEK',
  'EGP',
  'GBP',
  'GRD',
  'HKD',
  'HRK',
  'HUF',
  'IDR',
  'ILS',
  'INR',
  'ISK',
  'JPY',
  'KRW',
  'LTL',
  'LVL',
  'MAD',
  'MTL',
  'MXN',
  'MYR',
  'NOK',
  'NZD',
  'PHP',
  'PLN',
  'RON',
  'RUB',
  'SEK',
  'SGD',
  'SIT',
  'SKK',
  'THB',
  'TRY',
  'TWD',
  'USD',
  'ZAR',
];

export const processBudgetItem = async (
  inputBudgetItem: BudgetItem,
  projectEnd: string,
  projectEndPrecision: string
) => {
  if (!inputBudgetItem) return null;

  // check if it at least one is in EUR
  if (inputBudgetItem.currency === 'EUR') return inputBudgetItem;

  // ensure date is provided, as it's required
  if (!projectEnd) return inputBudgetItem;

  if (
    inputBudgetItem.currency &&
    inputBudgetItem.currency !== 'EUR' &&
    availableCurrencies.indexOf(inputBudgetItem.currency) >= 0
  ) {
    const projectEndDate = new Date(projectEnd);

    const euroValue = await getEuroValue(
      inputBudgetItem,
      projectEndDate,
      projectEndPrecision
    );

    if (euroValue) {
      const formattedEuroBudget = sanitizeBudgetItem({
        value: euroValue,
        currency: 'EUR',
        raw: `EUR ${euroValue}`,
        _original: {
          value: inputBudgetItem.value,
          currency: inputBudgetItem.currency,
          raw: inputBudgetItem.raw,
        },
      });

      // Return enriched record
      return formattedEuroBudget;
    }
  }

  return inputBudgetItem;
};

export default processBudgetItem;
