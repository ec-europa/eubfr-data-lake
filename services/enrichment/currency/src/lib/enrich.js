import { mergeRecords } from './merge';
// import { enrichCurrency } from '../plugins/currency';

export const enrich = async (record, existingRecord) => {
  /* if (!record.project_locations || record.project_locations.length === 0) {
    return record;
  } */

  const enrichedRecord = mergeRecords(record, existingRecord);

  // DO THE WORK
  console.log('ENRICH CURRENCY');

  return enrichedRecord;
};

export default enrich;
