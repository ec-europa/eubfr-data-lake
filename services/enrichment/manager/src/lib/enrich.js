import { mergeRecords } from './merge';
import { enrichFromCoordinates } from './plugins/coordinates';
import { enrichFromCountry } from './plugins/country';

export const enrich = async (record, existingRecord) => {
  const enrichedRecord = mergeRecords(record, existingRecord);

  const newLocations = (await Promise.all(
    enrichedRecord.project_locations.map(async loc => {
      // No information to work with, just exit
      if (!loc) {
        return null;
      }

      // Best scenario: we have coordinates
      if (loc.centroid) {
        // get nuts information from coordinates
        enrichFromCoordinates(loc);
      } else if (loc.address) {
        // get coordinates from address
        // if data added, call previous enrichFromCoordinates plugin here as well
      }

      // Less precision
      if (loc.nuts.length) {
        // get country code if not available yet
        // separate or only in previous step?
      }

      if (loc.country_code) {
        // find centroid's coordinates and add them
        enrichFromCountry(loc);
        // marked as enriched in order to avoid further enrichment based on low precision
        // eslint-disable-next-line
        loc.enriched = true;
      }

      return loc;
    })
  )).filter(loc => loc);

  // Update locations
  enrichedRecord.project_locations = newLocations;

  return enrichedRecord;
};

export default enrich;
