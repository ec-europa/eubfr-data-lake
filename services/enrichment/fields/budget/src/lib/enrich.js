/**
 * Convert foreign currencies into EUR usign the exchange rate at project end
 * https://webgate.ec.europa.eu/CITnet/jira/browse/EUBFR-184
 */

import { processBudgetItem } from './processBudgetItem';

export const enrich = async record => {
  if (!record) return null;

  const enrichedRecord = JSON.parse(JSON.stringify(record));

  // If project end date or the budget is not provided
  // or if '_original' is already present,
  // then skip enrichment
  if (
    !enrichedRecord.timeframe ||
    !enrichedRecord.timeframe.to ||
    !enrichedRecord.budget ||
    enrichedRecord.budget._original
  ) {
    return enrichedRecord;
  }

  const budgetFields = [
    'eu_contrib',
    'other_contrib',
    'private_fund',
    'public_fund',
    'total_cost',
  ];

  const enrichedBudgetFields = await Promise.all(
    budgetFields.map(async field =>
      processBudgetItem(
        enrichedRecord.budget[field],
        enrichedRecord.timeframe.to,
        enrichedRecord.timeframe.to_precision
      )
    )
  );

  // Enrich here
  enrichedBudgetFields.forEach((field, index) => {
    enrichedRecord.budget[budgetFields[index]] = field;
  });

  return enrichedRecord;
};

export default enrich;
