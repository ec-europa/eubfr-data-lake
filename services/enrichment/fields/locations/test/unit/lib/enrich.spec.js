/**
 * @jest-environment node
 */

import { enrich } from '../../../src/lib/enrich';
import existingRecord from './existingRecord.json';

describe('Enrichment module', () => {
  test('When empty objects: does not do anything', async () => {
    const result = await enrich({});
    expect(result).toMatchSnapshot();
  });

  test('When two records contain project_locations field and are to be merged', async () => {
    const result = await enrich(existingRecord);
    expect(result).toMatchSnapshot();
  });
});
