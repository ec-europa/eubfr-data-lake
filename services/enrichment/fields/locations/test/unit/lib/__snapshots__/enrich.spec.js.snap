// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enrichment module When empty objects: does not do anything 1`] = `Object {}`;

exports[`Enrichment module When two records contain project_locations field and are to be merged 1`] = `
Object {
  "action": "",
  "budget": Object {
    "eu_contrib": Object {
      "currency": "",
      "raw": "",
      "value": 0,
    },
    "funding_area": Array [],
    "mmf_heading": "",
    "other_contrib": Object {
      "currency": "",
      "raw": "",
      "value": 0,
    },
    "private_fund": Object {
      "currency": "",
      "raw": "",
      "value": 0,
    },
    "public_fund": Object {
      "currency": "",
      "raw": "",
      "value": 0,
    },
    "total_cost": Object {
      "currency": "",
      "raw": "",
      "value": 0,
    },
  },
  "call_year": "",
  "computed_key": "iati/45da84c2-a349-4610-a03a-85ad86d092fd.csv.ndjson",
  "created_by": "AWS:AROAIJA6MYF2IOUY6G5TU:chernka1622-ingestion-etl-iati-csv-parseCsv",
  "description": "bar",
  "ec_priorities": Array [],
  "last_modified": "2018-04-13T07:09:16.000Z",
  "media": Object {
    "cover_image": "",
    "video": "",
  },
  "programme_name": "",
  "project_id": "2b6cdeacbf83f07fb795b393f15df20f",
  "project_locations": Array [
    Object {
      "address": "",
      "centroid": Object {
        "lat": 50.6407351,
        "lon": 4.66696,
      },
      "country_code": "BE",
      "enriched": true,
      "location": Object {
        "coordinates": Array [
          4.66696,
          50.6407351,
        ],
        "type": "Point",
      },
      "nuts": Array [],
      "postal_code": "",
      "region": "235",
      "town": "",
    },
  ],
  "project_website": "",
  "related_links": Array [],
  "reporting_organisation": "European Commission - Development and Cooperation-EuropeAid",
  "results": Object {
    "available": "",
    "result": "",
  },
  "status": "",
  "sub_programme_name": "",
  "success_story": "",
  "themes": Array [],
  "third_parties": Array [],
  "timeframe": Object {
    "from": "2005-02-01T00:00:00.000Z",
    "to": "2005-10-31T00:00:00.000Z",
  },
  "title": "Foo",
  "type": Array [],
}
`;
